src/snowflake/connector/aio/_file_transfer_agent.py

```diff
@@ -0,0 +1,308 @@
+                m.get_callback = self._get_callback
+                m.get_azure_callback = self._get_azure_callback
+                m.get_callback_output_stream = self._get_callback_output_stream
+                m.show_progress_bar = self._show_progress_bar
+
+                # multichunk threshold
+                m.multipart_threshold = self._multipart_threshold
+
+        logger.debug(f"parallel=[{self._parallel}]")
+        if self._raise_put_get_error and not self._file_metadata:
+            Error.errorhandler_wrapper(
+                self._cursor.connection,
+                self._cursor,
+                OperationalError,
+                {
+                    "msg": "While getting file(s) there was an error: "
+                    "the file does not exist.",
+                    "errno": ER_FILE_NOT_EXISTS,
+                },
+            )
+        await self.transfer(self._file_metadata)
+
+        # turn enum to string, in order to have backward compatible interface
+
+        for result in self._results:
+            result.result_status = result.result_status.value
+
+    async def transfer(self, metas: list[SnowflakeFileMeta]) -> None:
+        files = [self._create_file_transfer_client(m) for m in metas]
+        is_upload = self._command_type == CMD_TYPE_UPLOAD
+        finish_upload_tasks = []
+        finish_download_tasks = []
+
+        async def preprocess_done_cb(
+            success: bool,
+            result: Any,
+            done_client: SnowflakeStorageClient,
+        ) -> None:
+            if not success:
+                logger.debug(f"Failed to prepare {done_client.meta.name}.")
+                if is_upload:
+                    await done_client.finish_upload()
+                    done_client.delete_client_data()
+                else:
+                    await done_client.finish_download()
+            elif done_client.meta.result_status == ResultStatus.SKIPPED:
+                # this case applies to upload only
+                return
+            else:
+                logger.debug(f"Finished preparing file {done_client.meta.name}")
+                tasks = []
+                for _chunk_id in range(done_client.num_of_chunks):
+                    task = (
+                        asyncio.create_task(done_client.upload_chunk(_chunk_id))
+                        if is_upload
+                        else asyncio.create_task(done_client.download_chunk(_chunk_id))
+                    )
+                    task.add_done_callback(
+                        lambda t, ta=task, dc=done_client, _chunk_id=_chunk_id: transfer_done_cb(
+                            ta, dc, _chunk_id
+                        )
+                    )
+                    tasks.append(task)
+                await asyncio.gather(*tasks)
+                if is_upload:
+                    await asyncio.gather(*finish_upload_tasks)
+                else:
+                    await asyncio.gather(*finish_download_tasks)
```

Comment:
I feel this chain of callback -- creating new asyncio task inside the callback is a bit unintuitive to understand and maintain, also I'm thinking if consolidating into coroutine could give some perf improvement.

let's log a jira to refactor this before we go GA?